### PR TITLE
Mailchimp Sync issue - Added check for valid status value

### PR DESCRIPTION
**Pull request by ujjwal-ab**
## Description of the change

> Added check for the possible status value and added one test case for that

Notion Link

[Ticket](https://www.notion.so/rudderstacks/Mailchimp-Sync-Issue-2ef987a31c2c4745bf17b323ba701f1c)

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: v0/destinations/mailchimp/transform.js [hunk 1]
```diff
@@ -100,6 +100,19 @@ async function responseBuilderSimple(payload, message, mailChimpConfig) {
   const basicAuth = Buffer.from(`apiKey:${mailChimpConfig.apiKey}`).toString(
     "base64"
   );
+  const validStatus = [
```
Comment (saikumarrs) [inline v0/destinations/mailchimp/transform.js:103]: This is more appropriate
```suggestion
  const validStatuses = [
```

File: v0/destinations/mailchimp/transform.js [hunk 2]
```diff
@@ -100,6 +100,19 @@ async function responseBuilderSimple(payload, message, mailChimpConfig) {
...
+    "subscribed",
+    "unsubscribed",
+    "cleaned",
+    "pending",
+    "transactional"
+  ];
+  if (payload.status && !validStatus.includes(payload.status)) {
+    throw new CustomError(
+      "The status does not contain any possible value",
```
Comment (saikumarrs) [inline v0/destinations/mailchimp/transform.js:112]: 1. The error message is incorrectly worded.
```suggestion
      "The status in the payload is invalid",
```
2. Just referring to `status` in the message might not make a whole lot of sense as this error message is user-facing. We should point out the field (that maps to `status`) from the source event as we seek corrective action from the users.

Comment (utsabc) [inline v0/destinations/mailchimp/transform.js:112]:
```suggestion
      "Invalid status must be one of [subscribed, unsubscribed, cleaned, pending, transactional]",
```

File: v0/destinations/mailchimp/transform.js (hunk 1 above)
Comment (utsabc) [inline v0/destinations/mailchimp/transform.js:103]: Can we move `validStatus` to config.js  and import it and use it here, make the code clean here

**Review by utsabc — CHANGES_REQUESTED** [review]

**Review by utsabc — DISMISSED** [review]

File: v0/destinations/mailchimp/transform.js (hunk 1 above)
Comment (ujjwal-ab) [reply v0/destinations/mailchimp/transform.js]: Updated that.

File: v0/destinations/mailchimp/transform.js (hunk 2 above)
Comment (ujjwal-ab) [reply v0/destinations/mailchimp/transform.js]: Updated the error message.

Comment (ujjwal-ab) [reply v0/destinations/mailchimp/transform.js]: Updated it.

File: v0/destinations/mailchimp/transform.js (hunk 1 above)
Comment (ujjwal-ab) [reply v0/destinations/mailchimp/transform.js]: Did that.

**Review by ItsSudip — DISMISSED** [review]

File: v0/destinations/mailchimp/transform.js (hunk 2 above)
Comment (saikumarrs) [reply v0/destinations/mailchimp/transform.js]: Point 2 is still not resolved.

File: v0/destinations/mailchimp/transform.js [hunk 3]
```diff
@@ -100,6 +101,12 @@ async function responseBuilderSimple(payload, message, mailChimpConfig) {
   const basicAuth = Buffer.from(`apiKey:${mailChimpConfig.apiKey}`).toString(
     "base64"
   );
+  if (payload.status && !validStatuses.includes(payload.status)) {
+    throw new CustomError(
+      "Invalid status must be one of [subscribed, unsubscribed, cleaned, pending, transactional]",
```
Comment (saikumarrs) [inline v0/destinations/mailchimp/transform.js:106]:
```suggestion
      "The status must be one of [subscribed, unsubscribed, cleaned, pending, transactional]",
```

Comment (saikumarrs) [inline v0/destinations/mailchimp/transform.js:106]: [Non-blocking comment] This is not maintainable as now the list of statuses is available in two places. We should avoid that.

**Review by saikumarrs — APPROVED** [review]

**Comment by ujjwal-ab** [comment]
Closing this PR due to conflicts and raising the new PR
 https://github.com/rudderlabs/rudder-transformer/pull/1044

**Pull request closed without merging**